### PR TITLE
sort attributes by frequency

### PR DESCRIPTION
**Pull request by alexlamsl**
This is a straight-forward implementation, with the following limitations/TODOs:
- [x] does not account for or propagate statistics to `processScripts`
- [x] does not omit removed attributes by other optimisations

### CODE REVIEW

**Comment by alexlamsl** [comment]
It does improve gzip ratios in `node benchmark.js`, which is nice to see.

**Comment by alexlamsl** [comment]
... oh and while we are sorting things, I've got bored looking for options, so I've sorted the list in `README.md` and `sample-cli-config-file.conf`.

File: src/htmlminifier.js [hunk 1]
```diff
@@ -662,6 +662,33 @@ function uniqueId(value) {
...
+    start: function(tag, attrs) {
+      var counter = counters[tag] || (counters[tag] = Object.create(null));
+      for (var i = 0; i < attrs.length; i++) {
+        var attr = attrs[i];
+        counter[attr.name] = (counter[attr.name] || 0) + 1;
+      }
+    }
+  });
+  return function(tag, attrs) {
+    if (attrs.length > 1) {
```
Comment (kangax) [inline src/htmlminifier.js]: Early return would be better here (to reduce depth)

File: src/htmlminifier.js [hunk 2]
```diff
@@ -662,6 +662,33 @@ function uniqueId(value) {
   return id;
 }
 
+function weightAttributes(value) {
```
Comment (kangax) [inline src/htmlminifier.js]: I think the name is a bit confusing for a function that creates another function. Maybe something like "makeAttrSorter"? Although that sounds awful too...

**Comment by kangax** [comment]
This is exciting. What are the actual gzip savings you saw from doing sorting? We should also check brotli.

File: src/htmlminifier.js (hunk 2 above)
Comment (alexlamsl) [inline src/htmlminifier.js]: `createAttributeSorter()`?

Comment (alexlamsl) [inline src/htmlminifier.js]: `createSortAttributeFn()`?

**Comment by alexlamsl** [comment]
Comments addressed - see if you like that new function name :sweat_smile: 

Here are the results with / without `sortAttributes`:

| gzip | After | Before | Delta |
| --- | --: | --: | --: |
| [HTMLMinifier page](https://github.com/kangax/html-minifier) | 8880 | 8917 | -37 |
| [NBC](http://www.nbc.com/) | 14937 | 14960 | -23 |
| [ES6 table](http://kangax.github.io/es5-compat-table/es6/) | 5265 | 5272 | -7 |
| [New York Times](http://www.nytimes.com/) | 25562 | 25571 | -9 |
| [Google](https://www.google.com/) | 39077 | 39079 | -2 |
| [MSN](http://www.msn.com/) | 37947 | 38006 | -59 |
| [Stackoverflow](http://stackoverflow.com/) | 30494 | 30490 | 4 |
| [Amazon](http://www.amazon.com/) | 49643 | 49703 | -60 |
| [Bootstrap CSS](http://getbootstrap.com/css/) | 38911 | 38930 | -19 |
| [Wikipedia](https://en.wikipedia.org/wiki/President_of_the_United_States) | 80333 | 80332 | 1 |
| [Eloquent Javascript](http://eloquentjavascript.net/1st_edition/print.html) | 192829 | 192559 | 270 |
| [ES6 draft](https://tc39.github.io/ecma262/) | 379722 | 379722 | 0 |

| lzma | After | Before | Delta |
| --- | --: | --: | --: |
| [HTMLMinifier page](https://github.com/kangax/html-minifier) | 8287 | 8325 | -38 |
| [NBC](http://www.nbc.com/) | 13743 | 13772 | -29 |
| [ES6 table](http://kangax.github.io/es5-compat-table/es6/) | 5013 | 5003 | 10 |
| [New York Times](http://www.nytimes.com/) | 23121 | 23159 | -38 |
| [Google](https://www.google.com/) | 35479 | 35492 | -13 |
| [MSN](http://www.msn.com/) | 34635 | 34688 | -53 |
| [Stackoverflow](http://stackoverflow.com/) | 27231 | 27207 | 24 |
| [Amazon](http://www.amazon.com/) | 44753 | 44805 | -52 |
| [Bootstrap CSS](http://getbootstrap.com/css/) | 34909 | 34966 | -57 |
| [Wikipedia](https://en.wikipedia.org/wiki/President_of_the_United_States) | 68987 | 68858 | 129 |
| [Eloquent Javascript](http://eloquentjavascript.net/1st_edition/print.html) | 169513 | 169491 | 22 |
| [ES6 draft](https://tc39.github.io/ecma262/) | 322638 | 322638 | 0 |

| brotli | After | Before | Delta |
| --- | --: | --: | --: |
| [HTMLMinifier page](https://github.com/kangax/html-minifier) | 7673 | 7654 | 19 |
| [NBC](http://www.nbc.com/) | 12979 | 13012 | -33 |
| [ES6 table](http://kangax.github.io/es5-compat-table/es6/) | 4350 | 4340 | 10 |
| [New York Times](http://www.nytimes.com/) | 22000 | 22021 | -21 |
| [Google](https://www.google.com/) | 35628 | 35590 | 38 |
| [MSN](http://www.msn.com/) | 33258 | 33295 | -37 |
| [Stackoverflow](http://stackoverflow.com/) | 25821 | 25868 | -47 |
| [Amazon](http://www.amazon.com/) | 42292 | 42280 | 12 |
| [Bootstrap CSS](http://getbootstrap.com/css/) | 31668 | 31678 | -10 |
| [Wikipedia](https://en.wikipedia.org/wiki/President_of_the_United_States) | 66061 | 66028 | -33 |
| [Eloquent Javascript](http://eloquentjavascript.net/1st_edition/print.html) | 160336 | 159950 | 386 |
| [ES6 draft](https://tc39.github.io/ecma262/) | 272776 | 272776 | 0 |

**Comment by kangax** [comment]
Nice! Other things to try — sorting inline styles? :) Although... I'm not sure we can reorder style values and have same behavior (in case one overwrites another and so on).

**Comment by kangax** [comment]
Also, one idea I had was to employ "adaptive hybrid" approach where we can test if sorted attributes result in smaller or larger output and chose the best branch. But that should probably be done a level higher than html-minifier...

**Comment by alexlamsl** [comment]
Inline styles are probably best left to `clean-css` to handle IMHO.

As for hybrid - we can cycle through all combinations of a given set of options and achieve that, and yes that should probably be done above `minify()`.

Back to this feature - I can account for removed tags & attributes if I run `minify()` once (with `sortAttributes` off) then run statistics on that result before running it again, but at a huge performance penalty.

Merging `processScripts` though I have a hinge on how to do it...

**Comment by alexlamsl** [comment]
On second thought, screw it - `sortAttributes` is off by default anyway, so taking a bit of a performance hit to make it work better when it is turned on isn't the end of the world.

**Comment by alexlamsl** [comment]
btw, there's #267 which I might give it a go after this...

**Comment by kangax** [comment]
@alexlamsl hey, were these results you pasted as a table done after sorting alphabetically or sorting by longest chain?

**Comment by alexlamsl** [comment]
@kangax I was just running `benchmark.js` on HEAD to see if I've screwed anything up :wink: 

`README.md` numbers are all current and up-to-date.

**Comment by kangax** [comment]
@alexlamsl ah, I was curious about deltas — if we get even better wins than the ones you pasted earlier.

**Comment by alexlamsl** [comment]
@kangax gotcha - here you go (`sortClassName` is off in both cases, obviously):

| gzip | Sort | Chain | Delta |
| --- | --: | --: | --: |
| [HTMLMinifier page](https://github.com/kangax/html-minifier) | 8880 | 8887 | 7 |
| [NBC](http://www.nbc.com/) | 14937 | 14937 | 0 |
| [ES6 table](http://kangax.github.io/es5-compat-table/es6/) | 5265 | 5265 | 0 |
| [New York Times](http://www.nytimes.com/) | 25562 | 25552 | -10 |
| [Google](https://www.google.com/) | 39077 | 39076 | -1 |
| [MSN](http://www.msn.com/) | 37947 | 37945 | -2 |
| [Stackoverflow](http://stackoverflow.com/) | 30494 | 30494 | 0 |
| [Amazon](http://www.amazon.com/) | 49643 | 49645 | 2 |
| [Bootstrap CSS](http://getbootstrap.com/css/) | 38911 | 38918 | 7 |
| [Wikipedia](https://en.wikipedia.org/wiki/President_of_the_United_States) | 80333 | 80332 | -1 |
| [Eloquent Javascript](http://eloquentjavascript.net/1st_edition/print.html) | 192829 | 192827 | -2 |
| [ES6 draft](https://tc39.github.io/ecma262/) | 379722 | 379722 | 0 |

| lzma | Sort | Chain | Delta |
| --- | --: | --: | --: |
| [HTMLMinifier page](https://github.com/kangax/html-minifier) | 8287 | 8296 | 9 |
| [NBC](http://www.nbc.com/) | 13743 | 13742 | -1 |
| [ES6 table](http://kangax.github.io/es5-compat-table/es6/) | 5013 | 5013 | 0 |
| [New York Times](http://www.nytimes.com/) | 23121 | 23104 | -17 |
| [Google](https://www.google.com/) | 35479 | 35475 | -4 |
| [MSN](http://www.msn.com/) | 34635 | 34629 | -6 |
| [Stackoverflow](http://stackoverflow.com/) | 27231 | 27239 | 8 |
| [Amazon](http://www.amazon.com/) | 44753 | 44742 | -11 |
| [Bootstrap CSS](http://getbootstrap.com/css/) | 34909 | 34900 | -9 |
| [Wikipedia](https://en.wikipedia.org/wiki/President_of_the_United_States) | 68987 | 68994 | 7 |
| [Eloquent Javascript](http://eloquentjavascript.net/1st_edition/print.html) | 169513 | 169643 | 130 |
| [ES6 draft](https://tc39.github.io/ecma262/) | 322638 | 322638 | 0 |

| brotli | Sort | Chain | Delta |
| --- | --: | --: | --: |
| [HTMLMinifier page](https://github.com/kangax/html-minifier) | 7673 | 7646 | -27 |
| [NBC](http://www.nbc.com/) | 12979 | 12966 | -13 |
| [ES6 table](http://kangax.github.io/es5-compat-table/es6/) | 4350 | 4350 | 0 |
| [New York Times](http://www.nytimes.com/) | 22000 | 21974 | -26 |
| [Google](https://www.google.com/) | 35628 | 35648 | 20 |
| [MSN](http://www.msn.com/) | 33258 | 33236 | -22 |
| [Stackoverflow](http://stackoverflow.com/) | 25821 | 25869 | 48 |
| [Amazon](http://www.amazon.com/) | 42292 | 42253 | -39 |
| [Bootstrap CSS](http://getbootstrap.com/css/) | 31668 | 31677 | 9 |
| [Wikipedia](https://en.wikipedia.org/wiki/President_of_the_United_States) | 66061 | 66043 | 18 |
| [Eloquent Javascript](http://eloquentjavascript.net/1st_edition/print.html) | 160336 | 160023 | -313 |
| [ES6 draft](https://tc39.github.io/ecma262/) | 272776 | 272776 | 0 |

I'd say we win a little bit more in general...

**Comment by kangax** [comment]
So... sometimes you get better win when they're sorted by most frequent, sometimes when by longest occurring chain, and sometimes without any optimizations?

**Comment by alexlamsl** [comment]
Yup - we are merely going with the most frequent winner, as with the usual case with compression schemes. :sweat_smile:

**Pull request merged**